org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -16,63 +16,84 @@ namespace MixedReality.Toolkit.Input
-                    if (!DetectedColliders.Contains(previouslyDetectedCollider) && previouslyDetectedCollider != null)
+                    for (int j = nearInteractables.Count - 1; j >= 0; j--)
                     {
-                        IXRProximityInteractable nearInteractionMode = previouslyDetectedCollider.GetComponent<IXRProximityInteractable>();
-                        if (nearInteractionMode != null)
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                previouslyDetectedCollider.GetComponent<IXRProximityInteractable>().OnProximityExited(new ProximityExitedEventArgs(this, previouslyDetectedCollider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Remove(previouslyDetectedCollider);
+                        xrProximityInteractable.OnProximityExited(new ProximityExitedEventArgs(previouslyDetectedCollider, nearInteractables[j]));
                     }
+                    previouslyDetectedColliders.Remove(previouslyDetectedCollider);
                 }
-                foreach (Collider collider in DetectedColliders)
+            }
+        }
+
+        /// <summary>
+        /// Call OnProximityEntered for all colliders that are currently detected but were not detected previously.  This
+        /// effectively triggers OnProximityEntered for all colliders that are newly detected.
+        /// </summary>
+        private void UpdateProximityEntered()
+        {
+            foreach (Collider collider in DetectedColliders)
+            {
+                if (!previouslyDetectedColliders.Contains(collider) &&
+                    InteractionManager.TryGetInteractableForCollider(collider, out IXRInteractable xrInteractable) &&
+                    xrInteractable is IXRProximityInteractable xrProximityInteractable)
                 {
-                    if (!previouslyDetectedColliders.Contains(collider))
+                    foreach (XRBaseInteractable xrBaseInteractable in nearInteractables)
                     {
-                        IXRProximityInteractable nearInteractionMode = collider.GetComponent<IXRProximityInteractable>();
-                        if (nearInteractionMode != null)
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                collider.GetComponent<IXRProximityInteractable>().OnProximityEntered(new ProximityEnteredEventArgs(this, collider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Add(collider);
+                        xrProximityInteractable.OnProximityEntered(new ProximityEnteredEventArgs(collider, xrBaseInteractable));
```

Comment:
You are still potentially calling `OnProximityEntered` twice on an interactable. if that interactable has multiple colliders.

I still think you should be storing `XRBaseInteractable` in `previouslyDetectedColliders`, instead of colliders.   I think this would help you prevent multiple calls to `OnProximityEntered`

Reply:
Actually, we need to store both XRBaseInteractable and Colliders because such combination uniquely identifies the hover-proximity event for individual buttons, otherwise some buttons won't reset their state to non-hover-proximity and would only react to Hover events (I tested it).  We had the Interactable+Collider earlier but we lost while implementing feedback.  I have refactored the code to retake such approach so that button frontplate rawimages react to hover+proximity events as it was originally behaving.  The refactoring is implemented in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/2de988906343015208b3b1a5f6069fa09347d3ea resolving this for another pass on the PR.